src/main/ank/1.6-simple-algebraic-data-types-Part1.md

```diff
@@ -0,0 +1,308 @@
+```
+```kotlin:ank:silent
+import arrow.core.Tuple3
+
+val startsWithSymbol: (Tuple3<String, String, Int>) -> Boolean =
+    {it.a.startsWith(it.b)}
+```
+................
+```Haskell
+data Element = Element { name :: String
+                       , symbol :: String
+                       , atomicNumber :: Int }
+```
+```kotlin:ank:silent
+data class Element(val name:String, val symbol:String, val atomicNumber:Int)
+```
+................
+```Haskell
+tupleToElem :: (String, String, Int) -> Element
+tupleToElem (n, s, a) = Element { name = n
+                                , symbol = s
+                                , atomicNumber = a }
+```
+```kotlin:ank:silent
+import arrow.core.Tuple3
+
+val tupleToElm: (Tuple3<String, String, Int>) -> Element = 
+    {Element(it.a, it.b, it.c)}
+```
+................
+```Haskell
+elemToTuple :: Element -> (String, String, Int) 
+elemToTuple e = (name e, symbol e, atomicNumber e)
+```
+```kotlin:ank:silent
+import arrow.core.Tuple3
+
+val elemToTuple: (Element) -> Tuple3<String, String, Int> =
+    {Tuple3(it.name, it.symbol, it.atomicNumber)}
+```
+................
+```Haskell
+atomicNumber :: Element -> Int
+```
+```kotlin
+val atomicNumber : (Element) -> Int
+```
+................
+```Haskell
+startsWithSymbol :: Element -> Bool 
+startsWithSymbol e = isPrefixOf (symbol e) (name e)
+```
+```kotlin:ank:silent
+val isPrefixOf = {prefix: String, str:String -> str.startsWith(prefix)}
+
+val startsWithSymbol: (Element) -> Boolean = {isPrefixOf(it.symbol, it.name)} 
+```
+................
+```Haskell
+startsWithSymbol e = symbol e `isPrefixOf` name e
+```
+```kotlin:ank:silent
+infix fun String.isPrefixOf(el:String) = isPrefixOf(this, el)
+
+val startsWithSymbol: (Element) -> Boolean = {it.symbol isPrefixOf it.name}
+```
+................
+```Haskell
+data Either a b = Left a | Right b
+```
+```kotlin:ank:silent
+import arrow.higherkind
+import arrow.core.EitherOf
+sealed class Either<out A, out B> : EitherOf<A, B> {
+    //… simplified
+    data class Left<out A>(val a: A) : Either<A, Nothing>() 
+    data class Right<out B>(val b: B) : Either<Nothing, B>()
+}
+```
+................
+```Haskell
+data OneOfThree a b c = Sinistral a | Medial b | Dextral c
+```
+```kotlin:ank:silent
+import arrow.generic.coproduct3.Coproduct3
+
+typealias OneOfThree<A, B, C> = Coproduct3<A, B, C>
+```
+................
+```Haskell
+Either a Void
+```
+```kotlin
+Either<A, Nothing>
+```
+................
+```Haskell
+data Color = Red | Green | Blue
+```
+```kotlin:ank:silent
+sealed class Color {
+    object Red : Color()
+    object Green : Color()
+    object Blue : Color()
+}
+```
+................
+```Haskell
+data Bool = True | False
+```
+```kotlin
+public class Boolean private constructor() : Comparable<Boolean> 
+```
+................
+```Haskell
+data Maybe a = Nothing | Just a
+```
+```kotlin:ank:silent
+import arrow.core.OptionOf
+
+sealed class Option<out A> : OptionOf<A>{
```

Comment:
remove `OptionOf` if you are not gonna place `@higherkind` which is fine for this example